tools/ir-generator/irclass.cpp

```diff
@@ -99,6 +99,35 @@ void exit_namespace(std::ostream &out, IrNamespace *ns) {
 
 ////////////////////////////////////////////////////////////////////////////////////
 
+/* sort class definitions so defs come before uses */
+void IrDefinitions::toposort() {
+    std::vector<IrElement *> sorted;
+    std::map<const IrClass *, IrClass *> classes;
+
+    // FIXME -- can't use auto here as it gives an error about using auto before it is deduced
+    std::function<void(const IrClass *)> visit = [&](const IrClass *cl) {
+        auto it = classes.find(cl);
+        if (it != classes.end()) {
+            auto *cl = it->second;
+            classes.erase(it);
+            visit(cl->concreteParent);
+            for (auto *p : cl->parentClasses) visit(p);
+            sorted.push_back(cl);
+        }
+    };
```

Comment:
It might be better to pull the function out and pass `classes` & `sorted` to it explicitly, instead of using this hack to create a recursive lambda (at the cost of dynamic invocation through `std::function`).

Reply:
Since the ir-generator runs in less than a second, I'm not too worried about the performance of it.

Reply:
I figured out a way to use a nested lambda to make this work auto properly.